### PR TITLE
add logerf(a, b)

### PR DESCRIPTION
**Pull request by cossio**
This PR adds a `logerf(a, b)` function, which computes `log(erf(b) - erf(a))` but maintaining accuracy.

Before this PR:

```julia
julia> log(erf(35) - erf(30))
-Inf
```

After this PR

```julia
julia> logerf(30, 35)
-903.9741171106439
```

Note that `logerf(a, b)` is useful to have, because otherwise one cannot access the probabilities in the tails of a normal distribution (which are zero unless one works in log-scale, as this example shows).

I've bumped the patch version here, so that we can tag a new release if we decide to merge this.

### CODE REVIEW

**Comment by cossio** [comment]
For example, this can be used to evaluate the `cdf` of a truncated normal in Distributions.jl. The current implementation fails far in the tail:

```julia
julia> using Distributions
julia> cdf(truncated(Normal(), 30, 35), 31)
NaN
```

But this PR, we can compute the correct CDF value, by doing something like this:

```
julia> logphi(a, b) = logerf(a / √2, b / √2);
julia> exp(logphi(30, 31) - logphi(30, 35))
0.9999999999999432
```

File: src/erf.jl [hunk 1]
```diff
@@ -506,3 +506,27 @@ end
...
+
+See also: [`erf(x,y)`](@ref SpecialFunctions.erf).
+"""
+function logerf(a::Real, b::Real)
+    if b > a > 0
+        return logerfc(a) + log1mexp(logerfc(b) - logerfc(a))
+    elseif a < b < 0
+        return logerfc(-b) + log1mexp(logerfc(-a) - logerfc(-b))
+    else
+        return log(erf(a, b))
```
Comment (stevengj) [inline src/erf.jl:526]: Won't this have the same problem  as #227?

**Comment by simonbyrne** [comment]
You might want to look at https://github.com/JuliaStats/Distributions.jl/blob/0c6cd50b435f0648eeb236d57162c66e704754d0/src/univariate/continuous/normal.jl#L139-L145

**Comment by cossio** [comment]
@simonbyrne That's just `logerfc` or `logerfcx` (https://github.com/JuliaMath/SpecialFunctions.jl/blob/fc960c66e56d230a83999d8fa76c2c5bba28441c/src/erf.jl#L466), which already exist here in SpecialFunctions. Perhaps `Distributions.jl` should use these to avoid code duplication.

You can use https://github.com/JuliaStats/Distributions.jl/blob/0c6cd50b435f0648eeb236d57162c66e704754d0/src/univariate/continuous/normal.jl#L139-L145 to compute the probability of a *finite* interval, just like I use `logerfc` and `logerfcx` here. But I think it'd be best to have a centralized implementation so that users don't have to rediscover this everytime.

**Comment by cossio** [comment]
If there are no objections, can we merge this?

File: src/erf.jl [hunk 2]
```diff
@@ -506,3 +506,29 @@ end
...
+    end
+end
+
+two(x::Number) = oftype(x, 2)
+two(::Type{T}) where {T} = convert(T, 2)
```
Comment (musm) [inline src/erf.jl:533]: Why is this needed?

Comment (cossio) [reply src/erf.jl]: Yes, I only need the first version.

File: src/erf.jl [hunk 3]
```diff
@@ -506,3 +506,28 @@ end
...
+        return logerfc(a) + log1mexp(logerfc(b) - logerfc(a))
+    elseif a < b < 0
+        return logerfc(-b) + log1mexp(logerfc(-a) - logerfc(-b))
+    else
+        return log(erf(a, b))
+    end
+end
+
+two(x::Number) = oftype(x, 2)
+log1mexp(x::Real) = x < -log(two(x)) ? log1p(-exp(x)) : log(-expm1(x))
```
Comment (musm) [inline src/erf.jl:533]:
```suggestion
log1mexp(x::Real) = x < -log(oftype(x,2)) ? log1p(-exp(x)) : log(-expm1(x))
```

Comment (cossio) [reply src/erf.jl:533]: Why is that an improvement? If it's a matter of style, I find `two(x)` more readable (similar to `zero(x)`, `one(x)`, etc.).

Comment (musm) [reply src/erf.jl:533]: `one` and `zero` are builtins and serve as the identity and zero element for various operations. This is just unnecessary since it's used in exactly one location and doesn't encompass  such meanings.

Comment (cossio) [reply src/erf.jl]: Well, I'm fine either way. So let's remove `two(x)`.

File: Project.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 name = "SpecialFunctions"
 uuid = "276daf66-3868-5448-9aa4-cd146d93841b"
-version = "0.10.2"
+version = "0.10.3"
```
Comment (musm) [inline Project.toml:3]: Since this is exporting a new function, don't we need a bump of the minor version

Comment (ararslan) [reply Project.toml:3]: Nope

Comment (musm) [reply Project.toml:3]: got it, thanks for letting me know

Comment (ararslan) [reply Project.toml:3]: It would be if the major version is >=1, but since we're still at 0, minor bumps are only for breaking changes, which new exports aren't considered to be. Note that this is a bit different from how things were in the METADATA days, where folks were encouraged to make minor version bumps for any new features.

**Comment by cossio** [comment]
Merge & tag?

**Comment by cossio** [comment]
Thanks!
